x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,728 @@
+    if (!isActionSkipped) {
+      isParamsUpdateSkipped = false;
+    }
+    return { ...acc, ...ruleParams } as RuleAlertType['params'];
+  }, existingRuleParams);
+
+  /* New rule schema migration logic
+  *
+  * If a rule's `prebuilt` param doesn't exist
+  * modify its params to add the prebuilt and customized fields
+  */
+  if (modifiedParams.prebuilt === undefined) {
+    modifiedParams.prebuilt = modifiedParams.immutable;
+    modifiedParams.customized = modifiedParams.customized ?? false;
+  }
+
+  return { modifiedParams, isParamsUpdateSkipped };
+};
+```
+
+The **Import Rules** - `POST /rules/_import` endpoint takes an `ndjson` file with rule info and parses it to create or update new rules. 
+
+In the `importRules` [method](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/import/import_rules_utils.ts), for each rule parsed from the `ndjson` file, we attempt to read its corresponding saved object by its `rule_id`.
+
+If the rule is not found, we create a rule a new rule with the params passed. If it is found, and the `overwrite` param is set to `true` in the body, we patch the existing rule (otherwise we throw and error).
+
+**In the case of creating a new rule:**
+
+- **if the rule is already migrated** (the `prebuilt` and `customized` are defined and passed in the input file): pass in those values unchanged to the `createRules` [method](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/crud/create_rules.ts) so that the rule is created with those values.
+- **if the rules has not yet been migrated**: pass in the `prebuilt` param to `createRules` with the value from the `immutable` param from the input file. If `immutable` is not defined, default to `false`. In both cases, set `customized` to `false` (since the rule cannot have been customized if it was not yet migrated).
+
+From [`x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/import/import_rules_utils.ts`](x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/import/import_rules_utils.ts):
+
+```ts
+/* If `rule_id` does not exist, create a new rule */
+if (rule == null) {
+
+  /* Set the prebuilt param to the parsed `prebuilt` param, or `immutable` for non-migrated rules 
+   * If neither are defined, `createRules` defaults its `prebuilt` param to `false`.
+   */
+  const isNewRulePrebuilt = parsedRule.prebuilt || parsedRule.immutable;
+
+  /* If rule is not migrated, set `customized` to false. Otherwise use the input value */
+  const isNewRuleCustomized = parsedRule.customized === undefined ? false : parsedRule.customized;
+
+  await createRules({
+    rulesClient,
+    params: {
+      ...parsedRule,
+      customized:  isNewRuleCustomized,
+      exceptions_list: [...exceptions],
+    },
+    prebuilt: isNewRulePrebuilt,
```

Comment:
The `createRules` function is like this because the parameters passed in the `CreateRuleOptions` object that it takes as parameter extend the `RuleCreateProps`, which don't have `prebuilt` or `immutable` as a field within it:

https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/crud/create_rules.ts

```ts
export interface CreateRulesOptions<T extends RuleCreateProps = RuleCreateProps> {
  rulesClient: RulesClient;
  params: T;
  id?: string;
  prebuilt?: boolean;
  // immutable?: boolean;
  defaultEnabled?: boolean;
  allowMissingConnectorSecrets?: boolean;
}

export const createRules = async ({
  rulesClient,
  params,
  id,
  // immutable = false,
  prebuilt = false,
  defaultEnabled = true,
  allowMissingConnectorSecrets,
}: CreateRulesOptions): Promise<SanitizedRule<RuleParams>> => {
  // const internalRule = convertCreateAPIToInternalSchema(params, immutable, defaultEnabled);
  const internalRule = convertCreateAPIToInternalSchema(params, prebuilt, defaultEnabled);
```

Since `createRules` is used to create new custom rules, create prebuilt rules (and also create rules when importing them), the `immutable` (or `prebuilt` field) has to be a separate parameter to the function, that is sometimes passed as `true`, and sometimes as `false` (and should set the value for both `immutable` and `prebuilt` in the future).

`customized` should be always set to `false` since we are creating a new rule.

Reply:
I added this explanation in the RFC where I explain the changes needed to the `createRules` method.

Reply:
Sorry, one thing I missed here but is even more important: the `params` object here passed to `createRule` extends from `RuleCreateProps`, which itself doesn't contain the `immutable` or `prebuilt` fields.

This makes sense: we don't want to have these parameters as part of the Rule creation API endpoint.